solver/src/oneinch_solver.rs

```diff
@@ -4,3 +4,232 @@
+use futures::future;
+use maplit::hashmap;
+use model::order::OrderKind;
+use rand::seq::SliceRandom as _;
+use std::fmt::{self, Display, Formatter};
+
+/// A GPv2 solver that matches GP **sell** orders to direct 1Inch swaps.
+#[derive(Debug)]
+pub struct OneInchSolver {
+    settlement_contract: GPv2Settlement,
+    client: OneInchClient,
+}
+
+impl OneInchSolver {
+    /// Creates a new 1Inch solver instance for specified settlement contract
+    /// instance.
+    pub fn new(settlement_contract: GPv2Settlement) -> Self {
+        Self {
+            settlement_contract,
+            client: Default::default(),
+        }
+    }
+
+    /// Settles a single sell order against a 1Inch swap.
+    async fn settle_order(&self, order: LimitOrder) -> Result<Settlement> {
+        debug_assert_eq!(
+            order.kind,
+            OrderKind::Sell,
+            "only sell orders should be passed to settle_order"
+        );
+
+        let spender = self.client.get_spender().await?;
+        let sell_token = ERC20::at(&self.web3(), order.sell_token);
+        let existing_allowance = sell_token
+            .allowance(self.settlement_contract.address(), spender.address)
+            .call()
+            .await?;
+
+        let swap = self
+            .client
+            .get_swap(SwapQuery {
+                from_token_address: order.sell_token,
+                to_token_address: order.buy_token,
+                amount: order.sell_amount,
+                from_address: self.settlement_contract.address(),
+                slippage: Slippage::basis_points(MAX_SLIPPAGE_BPS).unwrap(),
+                // Disable balance/allowance checks, as the settlement contract
+                // does not hold balances to traded tokens.
+                disable_estimate: Some(true),
+            })
+            .await?;
+
+        let mut settlement = Settlement::new(hashmap! {
+            order.sell_token => swap.to_token_amount,
+            order.buy_token => swap.from_token_amount,
+        });
+
+        settlement.with_liquidity(&order, order.sell_amount)?;
+
+        if existing_allowance < order.sell_amount {
+            settlement
+                .encoder
+                .append_to_execution_plan(Erc20ApproveInteraction {
+                    token: sell_token,
+                    spender: spender.address,
+                    amount: U256::MAX,
+                });
+        }
+        settlement.encoder.append_to_execution_plan(swap);
+
+        Ok(settlement)
+    }
+
+    fn web3(&self) -> DynWeb3 {
+        self.settlement_contract.raw_instance().web3()
+    }
+}
+
+impl Interaction for Swap {
+    fn encode(&self) -> Vec<EncodedInteraction> {
+        vec![(self.tx.to, self.tx.value, self.tx.data.clone())]
+    }
+}
+
+/// Maximum number of sell orders to consider for settlements.
+///
+/// This is mostly out of concern to avoid rate limiting and because 1Inch
+/// requests take a non-trivial amount of time.
+const MAX_SETTLEMENTS: usize = 5;
```

Comment:
Nit: Should this be moved to the top of the file?

Reply:
Personally I prefer it here, so its closer to the code that uses it. I don't feel very strongly about it. In general, the order things show up in a file isn't so important, AFAIK there is no "accepted style" for it, and I tend to prefer reading order in general.